crates/energy/src/battery.rs

```diff
@@ -0,0 +1,203 @@
+        }
+    }
+}
+
+
+
+impl Battery {
+    pub fn new(capacity: EnergyUnit, energy: EnergyUnit) -> Self {
+        let capacity = capacity.to_joules();
+        let energy = energy.to_joules();
+        debug_assert!(capacity.is_finite());
+        debug_assert!(capacity > 0.);
+        debug_assert!(energy.is_finite());
+        debug_assert!(energy >= 0.);
+        debug_assert!(energy <= capacity);
+
+        Self { capacity, energy }
+    }
+
+    /// The maximum capacity of the battery in kilojoules.
+    pub fn capacity(&self) -> f64 {
+        self.capacity
+    }
+
+    /// The current energy level of the battery in kilojoules.
+    pub fn energy(&self) -> f64 {
+        self.energy
+    }
+
+    /// Does the battery contain enough energy to move?
+    pub fn can_move(&self) -> bool {
+        self.energy >= MIN_MOVE_ENERGY .to_joules()
+    }
+
+    /// Does the battery contain enough energy to attack?
+    pub fn can_fire(&self) -> bool {
+        self.energy >= MIN_ATTACK_ENERGY.to_joules()
+    }
+
+    /// Does the battery contain enough energy to produce units?
+    pub fn can_produce(&self) -> bool {
+        self.energy >= MIN_FACTORY_ENERGY.to_joules()
+    }
+
+    /// Directly sets the energy level of the battery.
+    pub fn set_energy(&mut self, energy: f64) {
+        debug_assert!(energy.is_finite());
+        debug_assert!(energy >= 0.);
+        debug_assert!(energy <= self.capacity);
+
+        self.energy = energy;
+    }
+
+    /// Tries to discharge the battery by the given amount of energy.
+    ///
+    /// # Returns
+    ///
+    /// `true` if the battery was discharged, `false` otherwise.
+    pub fn try_discharge(&mut self, energy: EnergyUnit) -> bool {
+        let energy = energy.to_joules();
+        debug_assert!(energy.is_finite());
+        debug_assert!(energy >= 0.);
+
+        if energy >= self.energy {
+            return false;
+        }
+
+        self.change(-energy);
+        true
+    }
+
+    /// Directly changes the energy level of the battery by the given amount of energy.
+    pub fn change(&mut self, delta: f64) {
+        debug_assert!(delta.is_finite());
+
+        self.energy = (self.energy + delta).clamp(0., self.capacity);
+    }
+
+    /// Get fraction of energy remaining in the battery. (a number between 0 and 1)
+    pub fn energy_percentage(&self) -> f64 {
+        self.energy / self.capacity
+    }
+}
+
+/// Discharges the battery of a unit.
+///
+/// # Parameters
+///
+/// - `time`: The time since the last update in seconds.
+/// - `battery`: The battery.
+pub(crate) fn discharge_battery(time: Res<Time>, mut battery: Query<&mut Battery>) {
+    for mut battery in battery.iter_mut() {
+        let energy = battery.energy();
+        if energy == 0. {
+            continue;
+        }
+        let delta = time.delta_seconds();
+        let discharge = (DISCHARGE_RATE) * delta as f64;
+
+        battery.try_discharge(EnergyUnit::Joules(discharge));
```

Comment:
I am not a fun of this back and forth movement between pure f32 and EnergyUnit. It looks overengineered. I tried something similar but more powerful in `de_uom` where i.e. multiplication works out of the box so you don't have to do this transitions. I am still undecided whether it brings more benefits than complexity.

Please have a look at `de_uom` and either use that or stick to f32.